src/MemoryTracing/MemoryInfoProducer.cpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "MemoryTracing/MemoryInfoProducer.h"
+
+#include <absl/synchronization/mutex.h>
+#include <absl/time/time.h>
+
+#include <thread>
+
+#include "MemoryTracingUtils.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ThreadUtils.h"
+#include "capture.pb.h"
+
+namespace orbit_memory_tracing {
+
+using orbit_grpc_protos::SystemMemoryUsage;
+
+void MemoryInfoProducer::Start() {
+  SetExitRequested(false);
+  thread_ = std::make_unique<std::thread>(&MemoryInfoProducer::Run, this);
+}
+
+void MemoryInfoProducer::Stop() {
+  SetExitRequested(true);
+  if (thread_ != nullptr && thread_->joinable()) {
+    thread_->join();
+  }
+  thread_.reset();
+}
+
+void MemoryInfoProducer::SetExitRequested(bool status) {
+  absl::MutexLock lock(&exit_requested_mutex_);
+  *exit_requested_ = status;
+}
+
+void MemoryInfoProducer::Run() {
+  orbit_base::SetCurrentThreadName("MemInfoPr::Run");
+
+  CHECK(listener_ != nullptr);
+  ProduceSystemMemoryUsageAndSendToListener();
+}
+
+void MemoryInfoProducer::ProduceSystemMemoryUsageAndSendToListener() {
+  std::optional<SystemMemoryUsage> system_memory_usage;
+  absl::Time scheduled_time = absl::Now();
+  exit_requested_mutex_.ReaderLock();
```

Comment:
I feel like that with this read lock, we will not able to acquire a write lock and so a large `sampling_period_ns_`.

But when I try to remove these reader lock and unlock in line 49 and 58, and add the following method for reading `exit_requested_` in the current line 50 and line 56:
```cpp
bool MemoryInfoProducer::GetExitRequested() {
  exit_requested_mutex_.ReaderLock();
  bool exit_requested = *exit_requested_;
  exit_requested_mutex_.ReaderUnlock();
  return exit_requested;
}
```
I found the following error in the log file
```shell
[                OrbitService] [mutex.cc : 2410] RAW: thread should hold at least a read lock on Mutex
```

Reply:
yes - you probably want a regular lock here.. note that it will unlock on line 56, so the setter will be able to write at this point.

Reply:
Thanks so much! I got stuck on this for a long time!